datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -3029,6 +3048,24 @@ impl Aggregate {
     pub fn group_expr_len(&self) -> Result<usize> {
         grouping_set_expr_count(&self.group_expr)
     }
+
+    /// Returns the data type of the grouping id.
+    /// The grouping ID value is a bitmask where each set bit
+    /// indicates that the corresponding grouping expression is
+    /// null
+    pub fn grouping_id_type(group_exprs: usize) -> DataType {
+        if group_exprs <= 8 {
+            DataType::UInt8
+        } else if group_exprs <= 16 {
+            DataType::UInt16
+        } else if group_exprs <= 32 {
+            DataType::UInt32
+        } else {
+            DataType::UInt64
+        }
+    }
+
+    pub const INTERNAL_GROUPING_ID: &'static str = "__grouping_id";
```

Comment:
I recommend adding a high level description to this field explaining how the INTERNAL_GROUPING_ID is used with grouping sets (e.g explain the bitmask and have some examples), and then refer / link back to GROUPING_ID if needed.

That way if someone sees `__grouping_id` in their plan and `grep`s the code, they'll end up on the relevant documentation that gives the right overview

This could be done as a follow on PR as well

Reply:
Added the comment as part of rebasing the PR when solving a conflict.